File: packages/elixir-sdk/lib/treedx_sdk/pagination.ex
defmodule TreeDxSdk.Pagination do
  @moduledoc false
  defstruct items: [], next_cursor: nil, has_more: nil, cursor: nil, limit: nil

  def create_page(items, opts \\ []) do
    %__MODULE__{
      items: items,
      next_cursor: Keyword.get(opts, :next_cursor) || Keyword.get(opts, :nextCursor),
      has_more: Keyword.get(opts, :has_more) || Keyword.get(opts, :hasMore),
      cursor: Keyword.get(opts, :cursor),
      limit: Keyword.get(opts, :limit)
    }
  end

  def get_next_cursor(%__MODULE__{next_cursor: next_cursor}), do: next_cursor
  def page?(%__MODULE__{}), do: true
  def page?(_), do: false
end

File: packages/rust-sdk/tests/unit/pagination.rs
use treedx_sdk::TreeDxPage;
use treedx_sdk::pagination::{create_page, get_next_cursor};

#[test]
fn page_helpers_preserve_cursor_metadata() {
    let mut page = create_page(vec![1, 2]);
    page.next_cursor = Some("cursor-2".to_string());
    page.has_more = Some(true);
    assert_eq!(page.items, vec![1, 2]);
    assert_eq!(get_next_cursor(&page), Some("cursor-2"));
    assert_eq!(page.has_more, Some(true));

    let explicit = TreeDxPage {
        items: vec![3],
        next_cursor: None,
        has_more: Some(false),
        cursor: None,
        limit: Some(1),
    };
    assert_eq!(explicit.limit, Some(1));
}
